### PR TITLE
add integer casting support to VM

### PR DESCRIPTION
**Pull request by Parashurama**
This PR add support for casting between integers of same bitsize in VM (compile time and nimscript).
This allow to among other things to reinterpret signed integers as unsigned.

EDIT: PR now doesn't create a new opcode, but rather generate correct sequence of vm instruction to correctly cast signed <-> unsigned integers.
It also handle new behaviour of VM (VM int now emulate intsize used by target platform)

### CODE REVIEW

File: compiler/vmgen.nim [hunk 1]
```diff
@@ -1800,7 +1800,20 @@ proc gen(c: PCtx; n: PNode; dest: var TDest; flags: TGenFlags = {}) =
     if allowCast in c.features:
       genConv(c, n, n.sons[1], dest, opcCast)
     else:
-      globalError(n.info, errGenerated, "VM is not allowed to 'cast'")
+      let src = n.sons[1]
+
+      let dst_size = n.sons[0].sym.typ.size
```
Comment (Araq) [inline compiler/vmgen.nim]: that should be ``getSize(n[0].typ)``

File: compiler/vmgen.nim [hunk 2]
```diff
@@ -1800,7 +1800,20 @@ proc gen(c: PCtx; n: PNode; dest: var TDest; flags: TGenFlags = {}) =
...
+      let src_size = case src.kind
```
Comment (Araq) [inline compiler/vmgen.nim]: ``src.typ`` exists...

File: compiler/vm.nim [hunk 3]
```diff
@@ -1364,6 +1364,42 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
...
+      inc pc
+      let srctyp = c.types[c.code[pc].regBx - wordExcess]
+
+      proc opIntCast(dest: var TFullReg, src: TFullReg, desttyp, srctyp: PType) =
+        if dest.kind != rkInt:
+          reset(dest); dest.kind = rkInt
+        let val = src.intVal
+        case srctyp.kind
+        of tyUInt8, tyChar:
+          dest.intVal = if desttyp.kind in {tyUInt8, tyChar}: val
```
Comment (Araq) [inline compiler/vm.nim]: Types are static in Nim, the VM can eliminate casts from tyChar to tyChar, no need to check this at runtime in the VM.

File: compiler/vmgen.nim (hunk 2 above)
Comment (Parashurama) [inline compiler/vmgen.nim]: I'm not quite sure what you mean here. Are you saying I should just do
```let src_size = getSize(src.typ) ``` ?

Comment (Araq) [inline compiler/vmgen.nim]: Yes.

**Comment by dom96** [comment]
@Araq ping

File: tests/vm/tcastint.nim [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  doAssert(cast[uint32](I32) == U32)
+  doAssert(cast[uint64](I64A) == U64A)
+  doAssert(cast[uint64](I64B) == U64B)
+  doAssert(cast[int8](U8) == I8)
+  doAssert(cast[int16](U16) == I16)
+  doAssert(cast[int32](U32) == I32)
+  doAssert(cast[int64](U64A) == I64A)
+  doAssert(cast[int64](U64B) == I64B)
+  doAssert(cast[uint](IX) == UX) # uint is always 64 bits in VM
+  doAssert(cast[int](UX) == IX) # int is always 64 bits in VM
```
Comment (couven92) [inline tests/vm/tcastint.nim]: is `int` really always 64-bit in the VM? what if you compiled nim using `koch boot --cpu:i386`?

Comment (Parashurama) [inline tests/vm/tcastint.nim]: This is confirmed by @Araq. int is really always 64-bit in the VM (for the moment anyway). This might change in the future.

**Comment by data-man** [comment]
Next release? :)

**Pull request merged**